feat(quinn): Pass in `&mut self` for `AsyncUdpSocket`'s `poll_recv` and re-introduce `poll_send`

matheus23 opened the pull request:
This PR contains some changes to the `AsyncUdpSocket` trait. In short:
- `poll_recv` now takes `&mut self` instead of `&self`
- `try_send` is removed from `AsyncUdpSocket`
- `create_io_poller` and the `UdpPoller` trait are now replaced with `create_sender` and the `UdpSender` trait
- the `UdpSender` trait has a `try_send` as well as a `poll_send` function (that takes `self: Pin<&mut Self>`) instead of a `poll_writable` function

The final traits look like this:
```rs
pub trait AsyncUdpSocket: Send + Sync + Debug + 'static {
    fn create_sender(&self) -> Pin<Box<dyn UdpSender>>;

    fn poll_recv(
        &mut self,
        cx: &mut Context,
        bufs: &mut [IoSliceMut<'_>],
        meta: &mut [RecvMeta],
    ) -> Poll<io::Result<usize>>;

    fn local_addr(&self) -> io::Result<SocketAddr>;

    fn max_receive_segments(&self) -> usize;

    fn may_fragment(&self) -> bool;
}

pub trait UdpSender: Send + Sync + Debug + 'static {
    fn poll_send(
        self: Pin<&mut Self>,
        transmit: &Transmit,
        cx: &mut Context,
    ) -> Poll<io::Result<()>>;

    fn max_transmit_segments(&self) -> usize;
    
    fn try_send(self: Pin<&mut Self>, transmit: &Transmit) -> io::Result<()>;
}
```

## Motivation

The general idea is to encode the invariants that the runtime and socket implementations need to adhere to into rust's type system.
One of the harder things for me was to understand initially when I was working with quinn's `AsyncUdpSocket` was when you're allowed to overwrite the waker you're getting from `poll_recv` and when you can't, and similarly for the old `UdpPoller`.
Also, with iroh we had annoying issues with passing state from `poll_writable` to `try_send`. With a single `poll_send` that gives you (essentially) `&mut self` access, this gets *much* easier.

With all of these changes it's also a lot clearer when it's fine to overwrite wakers in your custom implementation. If `&mut self` is passed in from your `poll` function, then you *know* you can only be called from a single task that owns you at a time, thus you can safely overwrite old wakers (and even do so without any use of mutexes or fancy linked lists or anything like that).

## Performance

I've tried testing the performance of this on various systems (my laptop, dedicated bare metal servers, etc.). I wasn't able to determine any differences, but at the same time, I always got very inconsistent results (both on quinn `main` and this PR).
I'd really appreciate any help from someone who knows how to benchmark quinn effectively. There's some more things I can try, such as fiddling with the constants mentioned in https://github.com/quinn-rs/quinn/issues/1126, but other than that I'm out of ideas.
Looking at the code, this *should not* have a negative performance impact. I think there's technically 1 more Arc clone for initiating connections, but other than that it should result in pretty much the same amount of work/same sequence of syscalls.

Ralith commented on quinn/src/runtime.rs:
```diff
@@ -87,78 +76,161 @@ pub trait AsyncUdpSocket: Send + Sync + Debug + 'static {
+    ///
+    /// If sending in these cases fails, a well-behaved peer will re-try. Thus it's fine
+    /// if we drop datagrams sometimes with this function.
+    fn try_send(self: Pin<&mut Self>, transmit: &Transmit) -> io::Result<()>;
```
Would it be simpler to omit this and call `poll_send` with a no-op waker?

matheus23 replied:
Unfortunately, you don't have a dependency that provides a no-op waker yet. There's a `std::task::Waker::noop()`, but it's only stable since rust 1.85, which is far younger than your MSRV 1.71, so I've inlined the implementation.

Ralith commented on quinn/src/runtime.rs:
```diff
@@ -87,78 +76,161 @@ pub trait AsyncUdpSocket: Send + Sync + Debug + 'static {
+    ///
+    /// The `UdpSenderHelper` generic type parameters don't need to named, as it will be
+    /// used in its dyn-compatible form as a `Pin<Box<dyn UdpSender>>`.
+    pub struct UdpSenderHelper<Socket, MakeWritableFutFn, WritableFut> {
```
I don't love committing to this (now even more) complex and awkward utility as part of our public API. Can we split that out for separate discussion?

Ralith commented on quinn/src/runtime.rs:
```diff
@@ -266,7 +266,7 @@ mod tokio;
 #[cfg(feature = "runtime-tokio")]
 pub use self::tokio::TokioRuntime;
 
-#[cfg(feature = "async-io")]
```
Might be better to cfg out `async_io::UdpSocket` specifically? The `AsyncTimer` impl could, theoretically, be useful on its own, and so long as we have a freestanding async-io feature we might as well use it for that.

Might also be best to place this fix in a separate PR so we can merge it independently.

matheus23 replied:
In addition to that, even though the `async_io` module exists with only the `async-io` feature, it won't be exported, because the `pub use self::async_io::*` line below is feature-gated to the smol and async-std feature.
I don't think this makes sense, these two cfgs should just agree.

So. I can't *not* do changes here, as this breaks CI. But any other change doesn't seem sensical to me (and would be much bigger diffs). I'd argue this is the best fix?

Ralith replied:
> I don't really understand why it's needed for this PR, but not for main

That does seem strange. Maybe the feature testing is randomized? Maybe something in the toolchain updated since the last run on main?

> Given it causes a CI failure, I'm not 100% sure we can merge this independently.

I figured we'd merge it *first*, since with the proposed tweak it'd be non-breaking, and then this PR can be rebased and merged whenever we're ready to break.

> the pub use self::async_io::* line below is feature-gated to the smol and async-std feature.

We don't actually need to re-export anything for the trait impl on `Timer` to be available downstream. While it's a pretty niche hypothetical case where that's useful, masking it out is technically avoidable breakage.

matheus23 replied:
I have now rebased onto main after #2264

Ralith reviewed (changes requested):
Thanks! I think this is a good direction overall, though we may want to delay merging somewhat as it's breaking.

Ralith commented:
> Performance

I see no reason to expect that these changes present any performance hazard, and I'm satisfied with the smoke-testing you've done to check that assumption.

matheus23 commented:
> Thanks! I think this is a good direction overall, though we may want to delay merging somewhat as it's breaking.

Just to spell it out: I think I've addressed all reviewing concerns, and I think we're waiting only for the right moment to merge these breaking changes?

I realized I might be wrong though, so I'm double-checking now.

djc commented on quinn/src/runtime.rs:
```diff
@@ -87,77 +76,149 @@ pub trait AsyncUdpSocket: Send + Sync + Debug + 'static {
+impl<Socket, MakeWritableFutFn, WritableFut> Debug
+    for UdpSenderHelper<Socket, MakeWritableFutFn, WritableFut>
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```
Nit: import `fmt` please.

djc commented on quinn/src/runtime.rs:
```diff
@@ -87,77 +76,149 @@ pub trait AsyncUdpSocket: Send + Sync + Debug + 'static {
+            // If .writable() fails, propagate the error
+            result?;
+
+            let result = this.socket.try_send(transmit);
+
+            match result {
```
Nit: inline the `result` binding?

djc commented on quinn/src/runtime.rs:
```diff
@@ -87,77 +76,149 @@ pub trait AsyncUdpSocket: Send + Sync + Debug + 'static {
+                // `writable().await` call determines that the socket is indeed not writable and
+                // registers us for a wakeup, or the send succeeds if this really was just a
+                // transient failure.
+                Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => continue,
```
Nit: prefer to avoid `ref` in favor of match ergonomics.

djc commented on quinn/src/runtime/async_io.rs:
```diff
@@ -103,17 +103,27 @@ impl UdpSocket {
-            let socket = self.clone();
-            async move { socket.io.writable().await }
-        }))
+impl super::UdpSenderHelperSocket for Arc<UdpSocket> {
```
Nit: import `UdpSenderHelperSocket` directly.

matheus23 commented:
> I think it would be good to review the diff for that on top of the current state to compare the two approaches. Would be great if you can work on that.

I looked into this and sadly I think I was wrong earlier. This is impossible without performance implications, actually.
Even if the `UdpSenderHelper` were only used in one place, it would literally be impossible to change the abstract implementation into a concrete one, because you *need* the `WriteableFut` generic parameter, as you can't "name the future" that's returned from `tokio::net::UdpSocket::writable`. (The same goes for the `UdpPollHelper` on `main`)
The only option for naming that future is to box it, but that might mean worse performance.
The reason we don't need to box the future is that the generic argument is inferred to be an unnamed type in the compiler before we're forced to "name it" in syntax, by boxing the `UdpSenderHelper`/`UdpPollHelper` in its entirety in `AsyncUdpSocket::create_sender`/`AsyncUdpSocket::create_io_poller`.

matheus23 commented:
> For example, maybe `UdpSenderHelper` could own the `UdpSocketState` directly?

Unfortunately the implementation of `UdpSenderHelperSocket` references both the `UdpSocket` and the `UdpSocketState` in its `try_send` implementation, and `UdpSocket` is a different type between `runtime-smol` and `runtime-tokio`.

djc reviewed (approved):
Okay, let's get this in as is for now.
